src/arewefluentyet/aggregate-m2.py

```diff
@@ -0,0 +1,180 @@
+            return context
+    return None
+
+
+def extract_progress(raw_data):
+    contexts = []
+    entries = []
+
+    matches = re.finditer("DTD base/url: ([^,]+), (.*)", raw_data)
+    for match in matches:
+        if len(contexts) == 0 or contexts[-1]["origin"] != match.group(1):
+            if len(contexts) > 0:
+                contexts[-1]["to"] = match.start()
+
+            new_context = {
+                "from": match.start(),
+                "to": match.end(),
+                "origin": match.group(1),
+                "files": [match.group(2)]
+            }
+            contexts.append(new_context)
+        else:
+            contexts[-1]["files"].append(match.group(2))
+            contexts[-1]["to"] = match.end()
+
+    matches = re.finditer("== Entry ==(.*?)== Entry End ==", raw_data, re.DOTALL | re.MULTILINE)
+    for match in matches:
+        entry = {
+            "type": None,
+            "id": None,
+            "stack": None,
+            "file": None,
+            "stack": None,
+        }
+        stack = False
+        for line in match.group(1).replace("\\n", "\n").split("\n"):
+            if line.startswith("type: "):
+                entry["type"] = line[6:]
+            elif line.startswith("id: "):
+                entry["id"] = line[4:]
+            elif line.startswith("origin: "):
+                entry["stack"] = line[9:]
+            elif line.startswith("stack:"):
+                stack = True
+                entry["stack"] = ""
+            elif stack and len(line) > 0:
+                entry["stack"] += line + "\n"
+        if entry["stack"] is not None:
+            entry["stack"] = parse_stack(entry["stack"])
+
+        if entry["type"] == "dtd":
+            context = find_context(contexts, match.start())
+            if context:
+                entry["stack"] = context["origin"]
+                entry["file"] = context["files"][0]
+
+        entries.append(entry)
+
+    progress = {
+        "ftl": 0,
+        "dtd": 0,
+        "properties": 0
+    }
+    for entry in entries:
+        progress[entry["type"]] += 1
+
+    return (entries, progress)
+
+
+def get_bm_parent(mc_path, bm):
+    rev = subprocess.check_output([
+        "hg", "log",
+        "--cwd", mc_path,
+        "-r", "parents({})".format(bm),
+        "-T", "{node}",
+    ]).decode('utf-8')
+
+    return rev
+
+
+def apply_bm(mc_path, next_revision, bm):
+    parent = get_bm_parent(mc_path, bm)
+    if parent != next_revision:
+        subprocess.run([
+            "hg", "rebase", "--cwd", mc_path, "-s", bm, "-d", next_revision
+        ], check=True, stdout=subprocess.DEVNULL)
+
+
+def build_firefox(mc_path):
+    cmd = os.path.join(mc_path, "mach")
+    with open("firefox-build-log.txt", "w") as fp:
+        subprocess.run([cmd, "build"], stdout=fp, check=True)
+
+
+def collect_log(mc_path, log_name, next_date, next_revision):
+    data_dir = os.path.join(mc_path, "startup_log")
```

Comment:
It'd be nice to put the log file into the object dir, or maybe temp?

`./mach environment --format json` has info on the `topobjdir`, if that's what you'd go for.

Reply:
hmm, my worry is that this will get removed every time someone clobbers. I'm ok asking someone to point to the startup log dir that they maintain, if you prefer that over a dir in `m-c`. Would that work for you?